src/sparse.jl

```diff
@@ -0,0 +1,164 @@
+## Sparsity detector
+
+"""
+    AbstractSparsityDetector
+
+Abstract supertype for sparsity pattern detectors, defined for instance in [Symbolics.jl](https://github.com/JuliaSymbolics/Symbolics.jl).
+
+# Required methods
+
+- [`jacobian_sparsity`](@ref)
+- [`hessian_sparsity`](@ref)
+"""
+abstract type AbstractSparsityDetector end
+
+"""
+    jacobian_sparsity(f, x, sd::AbstractSparsityDetector)::AbstractMatrix{Bool}
+    jacobian_sparsity(f!, y, x, sd::AbstractSparsityDetector)::AbstractMatrix{Bool}
+
+Use detector `sd` to construct a (typically sparse) matrix `S` describing the pattern of nonzeroes in the Jacobian of `f` (resp. `f!`) applied at `x` (resp. `(y, x)`).
+"""
+function jacobian_sparsity end
+
+"""
+    hessian_sparsity(f, x, sd::AbstractSparsityDetector)::AbstractMatrix{Bool}
+
+Use detector `sd` to construct a (typically sparse) matrix `S` describing the pattern of nonzeroes in the Hessian of `f` applied at `x`.
+"""
+function hessian_sparsity end
+
+"""
+    NoSparsityDetector <: AbstractSparsityDetector
+
+Trivial sparsity detector, which always returns a full sparsity pattern (only ones, no zeroes).
+
+# See also
+
+- [`AbstractSparsityDetector`](@ref)
+"""
+struct NoSparsityDetector <: AbstractSparsityDetector end
+
+jacobian_sparsity(f, x, ::NoSparsityDetector) = ones(Bool, length(f(x)), length(x))
+jacobian_sparsity(f!, y, x, ::NoSparsityDetector) = ones(Bool, length(y), length(x))
+hessian_sparsity(f, x, ::NoSparsityDetector) = ones(Bool, length(x), length(x))
```

Comment:
```suggestion
jacobian_sparsity(f, x, ::NoSparsityDetector) = trues(Bool, length(f(x)), length(x))
jacobian_sparsity(f!, y, x, ::NoSparsityDetector) = trues(Bool, length(y), length(x))
hessian_sparsity(f, x, ::NoSparsityDetector) = trues(Bool, length(x), length(x))
```

Should it not assume Float64 and instead just be structural?

Reply:
I agree with the change but I don't understand the remark, ones(Bool) gives you the same as trues and has nothing to do with floats 🤔 ?

Reply:
I missed the Bool, though why use the extra space? A bitmatrix would be a more memory efficient form of the same structure and since it's not to be computed on and instead just be a structural representation I think it doesn't the accelerated read/write at the cost of size. Its main use would be conversion.

Reply:
good point, I didn't remember that `trues` outputs a `BitMatrix`, I'll make the switch